### DISCUSSION TITLE
Is there a reason to be concerned with memory usage when using the "UniqueObject" pattern?

### DISCUSSION THREAD

**Discussion by Alc-Alc** [sqlalchemy/sqlalchemy#11147] [Usage Questions] [answered]
Heya,

I am planning to use [UniqueObject](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject) and I noticed the "cache store" is a dictionary (`_unique_cache )`associated with the `Session` instance passed. If I were to use this mixin for a "large number of instances" <sup>*</sup> do I run a risk of facing memory issues due to holding the references in this cache? While it is recommended for the sessions to be short lived, I still have to ask if there are any downsides to storing these references?

My guess without knowing much about the implementation is that this would not be much of an overhead as the references that are stored in this cache would be the same ones that the `Session` instance would have when one does `session.add(SomeModel(...))`.

TL;DR - Other than the querying overhead, does "UniqueObject" pattern have any downsides with regards to memory consumption / garbage collection.

 <sup>*</sup> - I don't really have a quantification for what "large number" can be.

**Comment by zzzeek** [comment, marked as answer]
the cache is meant to be local to your Session and the work you are doing with it.    there are patterns with sessions where you are iterating through tens of thousands of objects, and that normally can be fine since the Session does not strongly reference objects - if you were doing that, and filling up the unique object cache, that might not be ideal.  it sort of depends on how you are using your session.  it's a recipe so that it's transparent, suggested solution to a particular problem.   if you are going to be upserting millions of rows you might want to use a different pattern like real upsert or working in ordered batches at a time.

**Reply by Alc-Alc** [reply]
Thanks for the quick reply :innocent:
